### PR TITLE
Refactor utils package

### PR DESCRIPTION
**Pull request by jmrossy**
### Description

- Reorganize code into separate files
- Move in utils from other packages
- Fix export structure
- Add multi-protocol address utils

### Drive-by changes

Lint fix in token package

### Related issues

Pre-req for https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/2503
Should merge before https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/2529

### Backward compatibility

No, some exports have been removed from the SDK and moved to the utils package

### CODE REVIEW

File: typescript/sdk/src/utils/filter.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { ChainMap } from '../types';
+
+export function filterByChains<T>(
```
Comment (jmrossy) [inline typescript/sdk/src/utils/filter.ts:3]: Moved here from sdk/src/utils/objects, the rest of which was moved to the utils package

File: typescript/utils/src/addresses.ts [hunk 2]
```diff
@@ -0,0 +1,240 @@
+import { PublicKey } from '@solana/web3.js';
```
Comment (jmrossy) [inline typescript/utils/src/addresses.ts:1]: Copied here from https://github.com/hyperlane-xyz/hyperlane-warp-ui-template/blob/main/src/utils/addresses.ts

File: typescript/utils/src/amount.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
+import { formatUnits, parseUnits } from '@ethersproject/units';
```
Comment (jmrossy) [inline typescript/utils/src/amount.ts:1]: Copied from https://github.com/hyperlane-xyz/hyperlane-warp-ui-template/blob/main/src/utils/amount.ts

File: solidity/test/isms/legacyMultisigIsm.test.ts [hunk 4]
```diff
@@ -57,7 +63,7 @@ describe('LegacyMultisigIsm', async () => {
   describe('#moduleType', () => {
     it('returns the correct type', async () => {
       expect(await multisigIsm.moduleType()).to.equal(
-        types.InterchainSecurityModuleType.MULTISIG,
```
Comment (jmrossy) [inline solidity/test/isms/legacyMultisigIsm.test.ts:60]: Directly exporting the contents of the utils package has 3 advantages:
1. Nicer DevX in editors
2. Allows for tree-shaking by bundlers
3. Avoids problem of people intentionally importing like this `@hyperlane-xyz/utils/dist/foo/bar/`

File: typescript/infra/package.json [hunk 5]
```diff
@@ -17,7 +17,7 @@
     "@nomiclabs/hardhat-etherscan": "^3.0.3",
     "@safe-global/api-kit": "^1.1.0",
     "@safe-global/protocol-kit": "^1.0.1",
-    "@solana/web3.js": "^1.77.3",
+    "@solana/web3.js": "^1.78.0",
```
Comment (jmrossy) [inline typescript/infra/package.json:20]: Just updating so we can use a consistent version everywhere, which may as well be newest

File: typescript/utils/src/arrays.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+interface Sliceable {
```
Comment (jmrossy) [inline typescript/utils/src/arrays.ts:1]: The rest of the files here in `utils/src/*` is code that was previously in the `utils/src/utils.ts` file or the SDK, just reorganized into specific files.

File: typescript/utils/src/addresses.ts [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+  const p2 = getAddressProtocolType(a2);
+  if (p1 !== p2) return false;
+  return routeAddressUtil(
+    (_a1) => areEvmAddressesEqual(_a1, a2),
+    (_a1) => areSolAddressesEqual(_a1, a2),
+    false,
+    a1,
+    p1,
+  );
+}
```
Comment (yorhodes) [inline typescript/utils/src/addresses.ts:128]: really don't love this naming haha :P

File: typescript/utils/src/addresses.ts [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+import { Address, HexString, ProtocolType } from './types';
+
+const EVM_ADDRESS_REGEX = /^0x[a-fA-F0-9]{40}$/;
+const SEALEVEL_ADDRESS_REGEX = /^[a-zA-Z0-9]{36,44}$/;
+
+const EVM_TX_HASH_REGEX = /^0x([A-Fa-f0-9]{64})$/;
+const SEALEVEL_TX_HASH_REGEX = /^[a-zA-Z1-9]{88}$/;
+
+const ZEROISH_ADDRESS_REGEX = /^(0x)?0*$/;
```
Comment (yorhodes) [inline typescript/utils/src/addresses.ts:12]: regex 🫣

File: typescript/utils/src/addresses.ts [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+
+export function shortenAddress(address: Address, capitalize?: boolean) {
+  if (!address) return '';
+  if (address.length < 8) return address;
+  const normalized = normalizeAddress(address);
+  const shortened =
+    normalized.substring(0, 5) +
+    '...' +
+    normalized.substring(normalized.length - 4);
+  return capitalize ? capitalizeAddress(shortened) : shortened;
```
Comment (yorhodes) [inline typescript/utils/src/addresses.ts:160]: is this like a pretty print thing?

File: typescript/utils/src/addresses.ts [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+export function capitalizeAddress(address: Address) {
+  if (address.startsWith('0x'))
+    return '0x' + address.substring(2).toUpperCase();
+  else return address.toUpperCase();
```
Comment (yorhodes) [inline typescript/utils/src/addresses.ts:166]: interesting
alternatively
`toUpperCase().replace('X, 'x')`

File: typescript/utils/src/amount.ts [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+import BigNumber from 'bignumber.js';
+
+const DEFAULT_MIN_ROUNDED_VALUE = 0.00001;
+const DEFAULT_DISPLAY_DECIMALS = 5;
+const DEFAULT_TOKEN_DECIMALS = 18;
+
+type NumberT = BigNumber.Value;
```
Comment (yorhodes) [inline typescript/utils/src/amount.ts:8]: what is this?

File: typescript/utils/src/amount.ts [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    if (!parsed || parsed.isNaN() || !parsed.isFinite()) return null;
+    else return parsed;
+  } catch (error) {
+    return null;
+  }
+}
+
+// Checks if an amount is equal of nearly equal to balance within a small margin of error
+// Necessary because amounts in the UI are often rounded
+export function areAmountsNearlyEqual(
```
Comment (yorhodes) [inline typescript/utils/src/amount.ts:78]: dont love naming here either

**Review by yorhodes — APPROVED** [review]

File: typescript/utils/src/addresses.ts (hunk 9 above)
Comment (jmrossy) [reply typescript/utils/src/addresses.ts:160]: Yep! For the UI

File: typescript/utils/src/addresses.ts (hunk 7 above)
Comment (jmrossy) [reply typescript/utils/src/addresses.ts:128]: Happy to rename, what would you prefer?

File: typescript/utils/src/addresses.ts (hunk 10 above)
Comment (jmrossy) [reply typescript/utils/src/addresses.ts:166]: Could screw up for base58 addresses like on sealevel

**Pull request merged**